### PR TITLE
feat: add simulations part 1

### LINKED ISSUE #294: add `x/module/simulation` for each module.
- [ ] re-enable simulations
- [ ] `x/mint`
- [ ] `x/claimsmanager`
- [ ] `x/epochs`
- [ ] `x/interchainquery`
- [ ] `x/interchainstaking`
- [ ] `x/participationrewards`
- [ ] `x/tokenfactory`

### PR DESCRIPTION
**Pull request by aljo242**
## 1. Summary
Part of #294
Fixes QCK-112
Fixes QCK-105
Fixes QCK-106
Fixes QCK-104

Add simulation testing back into the app and create a basic implementation of `x/mint` simulation.  `x/epochs` simulation already existed, so it is just added back in.

The remaining module simulations will be added piecemeal for each module and tracked in separate issues.

## 2.Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

## 3. Implementation details

- Add simulation manager to application in `/simulation`
- Add set of helper utilities for future simulation tests in `/simulation/simtypes` (adapted from https://github.com/osmosis-labs/osmosis/tree/main/simulation/simtypes)
- Set up simulation helper functions in `app/simulation`
- add `x/{moduleName}/simulation` for each of our custom modules
- add `simulation.yml` workflow which runs a lightweight version of `TestSimNonDeterminism`

## 4. How to test/use

`make test-sim-*` targets

## 5. Checklist

- [ ] Does the Readme need to be updated?

## 6. Limitations (optional)

Existing simulation "scenarios" need to be thoroughly checked.

## 7. Future Work (optional)

Add more sim functionality as needed

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -656,7 +657,6 @@ func NewQuicksilver(
 		gov.NewAppModule(appCodec, app.GovKeeper, app.AccountKeeper, app.BankKeeper),
 		slashing.NewAppModule(appCodec, app.SlashingKeeper, app.AccountKeeper, app.BankKeeper, app.StakingKeeper),
 		distr.NewAppModule(appCodec, app.DistrKeeper, app.AccountKeeper, app.BankKeeper, app.StakingKeeper),
-		mint.NewAppModule(appCodec, app.MintKeeper, app.AccountKeeper, app.BankKeeper),
```
Comment (joe-bowman) [inline app/app.go:659]: what's the rationale for moving this down? what is the impact?

Comment (joe-bowman) [reply app/app.go:659]: ah - because it is now under Quicksilver modules, not sdk modules?

**Review by joe-bowman — APPROVED** [review]
untested, but on the whole looks good. just the question around moving of the mint module, but think I answered that one myself!

File: app/app.go (hunk 1 above)
Comment (aljo242) [reply app/app.go:659]: Yeah, this doesn't impact functionality.  

I just moved mint to the  quicksilver modules for consistency.

**Review by joe-bowman — DISMISSED** [review]
lgtm

File: .github/workflows/simulation.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+name: simulate quicksilver
+
+on:
+  pull_request:
+  push:
+    paths-ignore:
+      - '*.md'
+    branches:
+      - main
```
Comment (ajansari95) [inline .github/workflows/simulation.yml:9]: simulation in general take a lot of ci minutes do we want to run it on every push? can this be a trigger that we can run once pr is ready to be merged with all sanity checks?

File: app/app.go [hunk 3]
```diff
@@ -804,27 +805,34 @@ func NewQuicksilver(
 
 	// // NOTE: this is not required apps that don't use the simulator for fuzz testing
```
Comment (ajansari95) [inline app/app.go:806]: can we remove this note now @joe-bowman

File: simulation/config.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	FlagBlockSizeValue          int
+	FlagLeanValue               bool
+	FlagCommitValue             bool
+	FlagOnOperationValue        bool // TODO: Remove in favor of binary search for invariant violation
+	FlagAllInvariantsValue      bool
+
+	FlagEnabledValue     bool
+	FlagVerboseValue     bool
+	FlagPeriodValue      uint
+	FlagGenesisTimeValue int64
```
Comment (ajansari95) [inline simulation/config.go:29]: just a nit, can this file be formatted, spacing doesn't seem to be consistent ?

**Review by ajansari95 — COMMENTED** [review]
looks good , some nits

File: .github/workflows/simulation.yml (hunk 2 above)
Comment (joe-bowman) [reply .github/workflows/simulation.yml:9]: I'm not sure how, but I've seen actions trigger on application of a 'ready-to-merge' type label. Is this easy to do?

Comment (ajansari95) [reply .github/workflows/simulation.yml:9]: i have also seen label based let me dig in a little

File: app/app.go (hunk 3 above)
Comment (joe-bowman) [inline app/app.go:806]:
```suggestion
```

File: app/app.go [hunk 5]
```diff
@@ -804,27 +805,34 @@ func NewQuicksilver(
...
 	// // transactions
```
Comment (joe-bowman) [inline app/app.go:807]:
```suggestion
```

File: app/app.go [hunk 6]
```diff
@@ -802,29 +803,34 @@ func NewQuicksilver(
...
+		epochs.NewAppModule(appCodec, app.EpochsKeeper),
+		interchainstakingModule,
+		interchainQueryModule,
+		participationrewardsModule,
+		airdropModule,
+		tokenfactory.NewAppModule(*app.TokenFactoryKeeper, app.AccountKeeper, app.BankKeeper),
+		mint.NewAppModule(appCodec, app.MintKeeper, app.AccountKeeper, app.BankKeeper),
+	)
 
 	// app.sm.RegisterStoreDecoders()
```
Comment (joe-bowman) [inline app/app.go:835]: @aljo242  - is this line not required?

**Comment by ajansari95** [comment]
@aljo242  merging `app.go` changes created conflicts as they are same. If we can resolve please

**Comment by aljo242** [comment]
@ajansari95 looks like the sim did not run.

Merge conflicts resolved

File: .github/workflows/simulation.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+name: simulate quicksilver
+
+on:
+  pull_request:
+    types: [ labeled ]
```
Comment (ajansari95) [inline .github/workflows/simulation.yml:5]: why is this workflow not running?

**Comment by ajansari95** [comment]
@joe-bowman  it dismiised your review as stale.

**Comment by joe-bowman** [comment]
sim didn't run? @ajansari95 @aljo242

**Comment by aljo242** [comment]
@ajansari95 @joe-bowman - action should be fixed now

**Comment by ajansari95** [comment]
> @ajansari95 @joe-bowman - action should be fixed now

yup looking in if you add labels on already existing PR will it work or not.

**Comment by aljo242** [comment]
Yeah I think this method should allow us to add the label at a later time - which is better

**Review by joe-bowman — APPROVED** [review]
lgtm

**Comment by joe-bowman** [comment]
@ajansari95 - your review was dismissed here ;)

**Pull request merged**